### ISSUE TITLE
[AIR 50.2][Android] `FREExtension::dispose()` and `FREContext::dispose()` not called

### ISSUE THREAD

**Issue by itlancer** [airsdk/Adobe-Runtime-Support#2614]
## Problem Description
Native extension `FREExtension::dispose()` and `FREContext::dispose()` methods not called with AIR 50.2.x anymore.
Not by direct AS3 `ExtensionContext::dispose()` call, not by application close or any.
So any native extensions cannot correctly dispose their resources.

Tested with latest AIR 50.2.2.3 and AIR 50.2.2.4 versions with multiple different Android devices with different OS versions with different applications and architectures.
Same problem in all cases.
There is no such problem with AIR 50.1.x and below.
Didn't test with other platforms yet.

May be related issues:
https://github.com/airsdk/Adobe-Runtime-Support/issues/2568
https://github.com/airsdk/Adobe-Runtime-Support/issues/2557
https://github.com/airsdk/Adobe-Runtime-Support/issues/2538

## Steps to Reproduce
1) Create any Android Java native extension with `FREContext::dispose()` method override.
```
package com.example; 
import java.util.HashMap; 
import java.util.Map; 
import android.util.Log; 
import com.adobe.fre.FREContext; 
import com.adobe.fre.FREFunction; 
 
public class MyNativeExtensionContext extends FREContext {
 
    public MyNativeExtensionContext() { 
        
    }
    
    @Override 
    public Map<String, FREFunction> getFunctions() { 
     Map<String, FREFunction> functionMap = new HashMap<String, FREFunction>();
	 return functionMap; 
    }
	
    @Override 
    public void dispose() {
	//This method will never called
        Log.i("MyNativeExtensionContext", "Dispose context"); 
    }
}
```
2) Create any AIR application that uses this native extension and somewhere call `ExtensionContext::dispose()` (on mouse click for example).
3) Launch application with any Android device and call `ExtensionContext::dispose()`. Or close application.

**Actual Result:**
Native extension `FREExtension::dispose()` and `FREContext::dispose()` methods not called.

**Expected Result:**
`FREExtension::dispose()` and `FREContext::dispose()` methods will be called.

## Known Workarounds
1) Use 50.1.x.
2) Create your own "dispose" function and call it directly as other native functions.

**Comment by pol2095** [comment]
I triedAir 0.2.2.4 in a Pixel 7 Pro and `FREContext::dispose()` is called.

**Comment by itlancer** [comment]
@pol2095, thanks. I made deeper investigation. It side effect using workers and application exit crash: https://github.com/airsdk/Adobe-Runtime-Support/issues/2571

**Comment by itlancer** [comment]
@ajwfrost 
Issue still exists. Tested with AIR SDK 50.2.3.1 with multiple different Android devices and architectures.
Issue can be reproduced only if application uses Workers and native extensions.
In such cases `FREContext::dispose()` not be called on application exit.
There is no such problem with AIR 50.1.x and below.
Also there is no such problem if no Workers started.

`runtimeInBackgroundThread>true</runtimeInBackgroundThread>` not used.

**Comment by ajwfrost** [comment]
Just to confirm on this one... the issue of the Java `FREContext::dispose()` not being called is now _only_ a problem just at application exit? Or is it still missing if you actually call `ExtensionContext.dispose()` manually?

We've not been able to reproduce the problem, we certainly get the call in Java if we call the AS3 code, and we also seem to get the call when all references to the ExtensionContext have been released. The only time we're not seeing the call is if the application quits and yet the ExtensionContext object was still being referenced.. so then, at application clean-up, the call into the `dispose()` method is not made because it would be unsafe if the player has been destroyed.

Would it be possible to get some AS3 code that demonstrates the issue here? I guess if this used to work in 50.1 but isn't working currently, then we should at least check what the difference might be - potentially if it was just caused by the introduction of the extension definitions into Workers, then we may still be able to correct the functionality..

thanks

**Comment by itlancer** [comment]
@ajwfrost

> Just to confirm on this one... the issue of the Java FREContext::dispose() not being called is now only a problem just at application exit? Or is it still missing if you actually call ExtensionContext.dispose() manually?

Both cases. But I agree with you about application exit (even if it works different with 50.1 and earlier in such case).
So, problem only with explicit `ExtensionContext.dispose()` call when application uses `Worker`.

Application sample with source code and native extension:
[java_ane_dispose_bug.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/12373941/java_ane_dispose_bug.zip)

Application:
```
package {
	import flash.display.Sprite;
	import com.test.javaane.JavaANE;
	import flash.events.MouseEvent;
	import flash.system.Worker;
	import flash.system.WorkerDomain;
	
	public class JavaANEDisposeBug extends Sprite {
		private var ane:JavaANE;
		private var worker:Worker;
		
		public function JavaANEDisposeBug() {
			if (Worker.current.isPrimordial) {
				//Create ANE
				ane = new JavaANE();
				
				worker = WorkerDomain.current.createWorker(this.loaderInfo.bytes, true);
				worker.start();
				
				stage.addEventListener(MouseEvent.CLICK, click);
			} else {
				trace("worker");
			}
		}
		
		public function click(e:MouseEvent):void {
			trace("click");
			ane.dispose();//This line should dispose native extension
		}
	}
}
```

AS3 part of native extension:
```
package com.test.javaane {
	import flash.external.ExtensionContext;
	
	public class JavaANE {
		private var _context:ExtensionContext;
		
		public function JavaANE() {
			_context = ExtensionContext.createExtensionContext("com.test.javaane", "");
		}
		
		public function dispose():void {
			_context.dispose();//This line should dispose native extension
			_context = null;
		}
	}
}
```

Java part of native extension:
```
package com.test.javaane;

import android.util.Log;

import com.adobe.fre.FREContext;
import com.adobe.fre.FREFunction;

import java.util.HashMap;
import java.util.Map;

public class JavaANEContext extends FREContext {

    public JavaANEContext() {

    }

	@Override
	public Map<String, FREFunction> getFunctions() {
		final Map<String, FREFunction> functions = new HashMap<>();
		return functions;
	}

	@Override
	public void dispose() {
		//This method not called if Worker used
		Log.i("JavaANEContext", "Dispose context");
	}
}
```

Tested again with latest AIR 50.2.3.3.
I can upload sample APK if necessary.

**Comment by ajwfrost** [comment]
Thanks .. we can reproduce with that. Really not sure why our own test case didn't have the same issue though..! We can see what's happening - the secondary worker is loading in the SWF definition for the extension (due to that change we made recently) but there's a state change that happens as part of that, which puts the extension back into an uninitialised state. The primary thread had gone beyond this state though, so it's then getting confused when it determines whether or not the dispose method actually needs passing through to the native layer..

**Comment by itlancer** [comment]
Fixed with latest AIR SDK 50.2.3.4.
Thanks!
